File: Exercise/Very Hard/03. Max Sum Longest Path in DAG/Code/Max Sum Longest Path in DAG.java
import java.util.*;

public class LongestPath {
    static final int NEG_INF = Integer.MIN_VALUE;

    static class Edge {
        int v, w;

        Edge(int v, int w) {
            this.v = v;
            this.w = w;
        }
    }

    public static List<Integer> findLongestPath(int n, List<List<Edge>> graph, int start) {
        int[] distance = new int[n];
        Arrays.fill(distance, NEG_INF);
        distance[start] = 0;
        List<Integer> topoOrder = new ArrayList<>();
        int[] indegree = new int[n];

        for (int u = 0; u < n; u++) {
            for (Edge edge : graph.get(u)) {
                indegree[edge.v]++;
            }
        }

        Deque<Integer> zeroIndegree = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                zeroIndegree.add(i);
            }
        }

        while (!zeroIndegree.isEmpty()) {
            int node = zeroIndegree.poll();
            topoOrder.add(node);
            for (Edge edge : graph.get(node)) {
                indegree[edge.v]--;
                if (indegree[edge.v] == 0) {
                    zeroIndegree.add(edge.v);
                }
            }
        }

        for (int node : topoOrder) {
            if (distance[node] != NEG_INF) {
                for (Edge edge : graph.get(node)) {
                    if (distance[edge.v] < distance[node] + edge.w) {
                        distance[edge.v] = distance[node] + edge.w;
                    }
                }
            }
        }

        int endNode = 0;
        for (int i = 1; i < n; i++) {
            if (distance[i] > distance[endNode]) {
                endNode = i;
            }
        }

        List<Integer> path = new ArrayList<>();
        int currentNode = endNode;
        while (currentNode != start) {
            path.add(currentNode);
            for (Edge edge : graph.get(currentNode)) {
                if (distance[currentNode] == distance[edge.v] + edge.w) {
                    currentNode = edge.v;
                    break;
                }
            }
        }
        path.add(start);
        Collections.reverse(path);

        return path;
    }

    public static void main(String[] args) {
        int n = 5;
        List<List<Edge>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        graph.get(0).add(new Edge(1, 5));
        graph.get(0).add(new Edge(2, 3));
        graph.get(1).add(new Edge(3, 6));
        graph.get(1).add(new Edge(2, 2));
        graph.get(2).add(new Edge(4, 4));
        graph.get(3).add(new Edge(4, -1));

        int start = 0;
        List<Integer> path = findLongestPath(n, graph, start);

        for (int node : path) {
            System.out.print(node + " ");
        }
        System.out.println();
    }
}

File: Exercise/Very Hard/03. Max Sum Longest Path in DAG/Code/Max Sum Longest Path in DAG.py
from collections import defaultdict, deque


def find_longest_path(n, edges, start):
    graph = defaultdict(list)
    for u, v, w in edges:
        graph[u].append((v, w))

    distance = [-float('inf')] * n
    distance[start] = 0
    topo_order = []
    indegree = [0] * n

    for u in range(n):
        for v, w in graph[u]:
            indegree[v] += 1

    zero_indegree = deque([i for i in range(n) if indegree[i] == 0])
    while zero_indegree:
        node = zero_indegree.popleft()
        topo_order.append(node)
        for neighbor, weight in graph[node]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                zero_indegree.append(neighbor)

    for node in topo_order:
        if distance[node] != -float('inf'):
            for neighbor, weight in graph[node]:
                if distance[neighbor] < distance[node] + weight:
                    distance[neighbor] = distance[node] + weight

    end_node = distance.index(max(distance))
    path = []
    current_node = end_node

    while current_node != start:
        path.append(current_node)
        for neighbor, weight in graph[current_node]:
            if distance[current_node] == distance[neighbor] + weight:
                current_node = neighbor
                break
    path.append(start)
    path.reverse()

    return path, distance[end_node]


def main():
    import sys
    input = sys.stdin.read
    data = input().split()

    n = int(data[0])
    edges = []
    index = 1
    for i in range(n):
        u = int(data[index]) - 1
        index += 1
        while index < len(data) and data[index] != '-1':
            v = int(data[index]) - 1
            index += 1
            w = int(data[index])
            index += 1
            edges.append((u, v, w))
        index += 1

    start = int(data[index]) - 1

    path, max_sum = find_longest_path(n, edges, start)
    print(" -> ".join(map(str, [x + 1 for x in path])))
    print("Sum:", max_sum)


if __name__ == "__main__":
    main()
